src/routes/auth.rs

```diff
@@ -129,78 +129,82 @@ pub async fn auth_callback(
     let mut transaction = client.begin().await?;
     let state_id = parse_base62(&*info.state)?;
 
-    let result = sqlx::query!(
+    let result_option = sqlx::query!(
         "
             SELECT url,expires FROM states
             WHERE id = $1
             ",
         state_id as i64
     )
-    .fetch_one(&mut *transaction)
+    .fetch_optional(&mut *transaction)
     .await?;
 
-    let now = Utc::now();
-    let duration = result.expires.signed_duration_since(now);
+    if let Some(result) = result_option {
```

Comment:
Invert that condition, you can make the code way more readable by testing for the contrary, and returning an error early.